### PR TITLE
Add limits for replicated fetches/sends bandwidth

### LINKED ISSUE #1821: Throttle recovery speed
We're in a process of migration from older hardware under ClickHouse to the newer generation. 

Older machines have 12x6T disks, 128GB RAM and 2x10G NICs, newer machines have 12x10T disks, 256GB RAM and 2x25G NICs. Dataset per replica is around 35TiB. Each shard is 3 replicas.

Our process is:

1. Stop one replica from shard.
2. Clear it from zookeeper.
3. Remove it from cluster topology (znode update for `remote_servers`).
4. Add new replica to cluster topology.
5. Start new replica and let it replicate all the data from peers.

The issue we're seeing is that source replicas saturate disks, starving user queries and merges.

It takes ~7h to replicate full dataset, below are the graphs for 12h around that time:

image

image

image

image

image

Source peer:

image

image

Target peer:

image

image

Naturally, source peers are not great at IO in the first place (that's why we're upgrading), but having 7h of degraded service is not great. It'd be nice to be able to set recovery speed, so it doesn't starve  other activities like user queries and merges.

Moreover, max number of parts in partition quickly reaches the max (we set to 1000) on the target replica, where inserts are throttled, which doesn't make things any better. With throttled recovery at lower speed but with higher duration, this will probably be even longer period. Maybe we should split threads that do merges and threads that do replication, it seems like whole pool is busy just replicating.

It is also possible that we're just doing it wrong, then it'd be great to have a guide describing the process.

cc @vavrusa, @dqminh, @bocharov

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add two Replicated*MergeTree settings: `max_replicated_fetches_network_bandwidth` and `max_replicated_sends_network_bandwidth` which allows to limit maximum speed of replicated fetches/sends for table. Add two server-wide settings (in `default` user profile): `max_replicated_fetches_network_bandwidth_for_server` and `max_replicated_sends_network_bandwidth_for_server` which limit maximum speed of replication for all tables. The settings are not followed perfectly accurately. Turned off by default. Fixes #1821.

### CODE REVIEW

**Comment by alesapin** [comment]
```
E               UnicodeEncodeError: 'ascii' codec can't encode characters in position 7-12: ordinal not in range(128)```

File: src/Common/Throttler.h [hunk 1]
```diff
@@ -97,6 +102,19 @@ class Throttler
 
         count = 0;
         watch.reset();
+        accumulated_sleep = 0;
+    }
+
+    /// Is throttler already accumulated some sleep time and throttling.
+    bool isThrottling() const
```
Comment (nikitamikhaylov) [inline src/Common/Throttler.h:109]: For me this method is unneeded. It is just like `size()` method of some concurrent data structure. This method won't tell you any needful information. And knowledge about throttling goes into other classes, so we have stronger dependencies.

File: src/Storages/MergeTree/DataPartsExchange.h [hunk 2]
```diff
@@ -51,7 +54,7 @@ class Service final : public InterserverIOEndpoint
 
     /// StorageReplicatedMergeTree::shutdown() waits for all parts exchange handlers to finish,
     /// so Service will never access dangling reference to storage
-    MergeTreeData & data;
+    StorageReplicatedMergeTree & data;
```
Comment (nikitamikhaylov) [inline src/Storages/MergeTree/DataPartsExchange.h:57]:
```suggestion
    StorageReplicatedMergeTree & storage;
```

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by alesapin** [comment]
We have to reset the throttler when there were not active fetches/sends. Otherwise, it will accumulate a huge wait time and will never throttle.

File: src/Common/Throttler.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        new_count = count;
+        prev_ns = now;
+    }
+
+    if (limit && new_count > limit)
+        throw Exception(limit_exceeded_exception_message + std::string(" Maximum: ") + toString(limit), ErrorCodes::LIMIT_EXCEEDED);
+
+    if (max_speed && current_speed > max_speed)
+    {
+        /// If we was too fast then we have to sleep util our smoothed speed became <= max_speed
```
Comment (nikitamikhaylov) [inline src/Common/Throttler.cpp:68]:
```suggestion
        /// If we was too fast then we have to sleep until our smoothed speed became <= max_speed
```

Comment (alesapin) [reply src/Common/Throttler.cpp:68]: I'll commit after checks and merge.

File: programs/client/Suggest.cpp [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include <Core/Settings.h>
 #include <Columns/ColumnString.h>
 #include <Common/typeid_cast.h>
+#include <sstream>
```
Comment (nikitamikhaylov) [inline programs/client/Suggest.cpp:6]:
```suggestion

```
I got rid from `std::stringstream` in this file #24882

**Comment by alesapin** [comment]
Flaky tests, need to investigate :(

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-10285.

**Pull request merged**